azkaban-common/src/main/java/azkaban/version/JdbcVersionSetLoader.java

```diff
@@ -0,0 +1,237 @@
+/*
+ * Copyright 2020 LinkedIn Corp.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package azkaban.version;
+
+import azkaban.db.DatabaseOperator;
+import azkaban.db.SQLTransaction;
+import java.io.IOException;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import org.apache.commons.dbutils.ResultSetHandler;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A JDBC based implementation for loading VersionSet from the backed db. This class keeps the copy
+ * of entries in the version_set table locally as well to reduce the number of queries to the db.
+ * <p>
+ * This is a singleton class to be instantiated using dependency injection.
+ */
+@Singleton
+public class JdbcVersionSetLoader implements VersionSetLoader {
+
+  private static final Logger logger = LoggerFactory.getLogger(JdbcVersionSetLoader.class);
+
+  private static final String SELECT_VSET_FROM_ID = "SELECT vset.id, vset.md5, vset.json FROM "
+      + "version_set vset WHERE vset.id=?";
+  private static final String SELECT_VSET_FROM_MD5 = "SELECT vset.id, vset.md5, vset.json FROM "
+      + "version_set vset WHERE vset.md5=?";
+  private static final String SELECT_ALL_VSET = "SELECT vset.id, vset.md5, vset.json FROM "
+      + "version_set vset";
+  private static final String INSERT_VSET = "INSERT INTO version_set (md5, json) values (?,?)";
+  private static final String DELETE_VSET = "DELETE FROM version_set vset WHERE vset.md5=?";
+
+  private final DatabaseOperator dbOperator;
+  private final Map<String, VersionSet> md5ToVersionSet = new HashMap<>();
+  private final Map<Integer, VersionSet> idToVersionSet = new HashMap<>();
+
+  /**
+   * This constructor should be instantiated using dependency injection only to make sure that that
+   * class remains Singleton.
+   *
+   * @param dbOperator Interface to perform db related operations.
+   */
+  @Inject
+  public JdbcVersionSetLoader(DatabaseOperator dbOperator) throws IOException {
+    this.dbOperator = dbOperator;
+    for (VersionSet versionSet : fetchAllVersionSets()) {
```

Comment:
As this is a Singleton class, fetching all the version set will happen only once during instance creation. We can improve on the caching part by using guava cache suppliers. Suppliers provide memoiseWithExpiration method, using which we can trigger db callback at specified duration say 1 hr or so. This will ensure that cache is updated at specified interval.

Reply:
I think this comment can be ignored as same class is populating the cache as well. I believe from nowhere else db will be populated.

Reply:
Yes, the same class will update the local maps as well as the db every time. Hence only once it needs to be loaded from the db.